pkg/api/htpasswd_test.go

```diff
@@ -0,0 +1,183 @@
+package api_test
+
+import (
+	"os"
+	"strings"
+	"testing"
+
+	. "github.com/smartystreets/goconvey/convey"
+	"golang.org/x/crypto/bcrypt"
+
+	zerr "zotregistry.io/zot/errors"
+	"zotregistry.io/zot/pkg/api"
+	test "zotregistry.io/zot/pkg/test/common"
+)
+
+func TestHtpasswdClient_ChangePassword(t *testing.T) {
+	Convey("test htpasswd client change oldPassword", t, func() {
+		username, _ := test.GenerateRandomString()
+		oldPassword, _ := test.GenerateRandomString()
+		newPassword, _ := test.GenerateRandomString()
+
+		htpasswdPath := test.MakeHtpasswdFileFromString(test.GetCredString(username, oldPassword))
+		defer os.Remove(htpasswdPath)
+
+		client := api.NewHtpasswdClient(htpasswdPath)
+		So(client.Init(), ShouldBeNil)
+
+		Convey("change for non-existing login", func() {
+			err := client.ChangePassword("non-existing", "old_password", "new_password")
+			So(err, ShouldEqual, zerr.ErrUserIsNotFound)
+		})
+
+		Convey("change with wrong old oldPassword", func() {
+			err := client.ChangePassword(username, "wrong_password", "new_password")
+			So(err, ShouldEqual, zerr.ErrOldPasswordIsWrong)
+
+			passphrase, ok := client.Get(username)
+			So(ok, ShouldBeTrue)
+			So(bcrypt.CompareHashAndPassword([]byte(passphrase), []byte(oldPassword)), ShouldBeNil)
+		})
+
+		Convey("change with empty new oldPassword", func() {
+			err := client.ChangePassword(username, oldPassword, "")
+			So(err, ShouldEqual, zerr.ErrPasswordIsEmpty)
+
+			passphrase, ok := client.Get(username)
+			So(ok, ShouldBeTrue)
+			So(bcrypt.CompareHashAndPassword([]byte(passphrase), []byte(oldPassword)), ShouldBeNil)
+		})
+
```

Comment:
Maybe test with empty user. Although we should validate the user in routes.go before reaching this logic.